src/get.rs

```diff
@@ -229,195 +175,772 @@ fn is_same_subnet(addr: &SocketAddr, interfaces: &[Interface]) -> bool {
+        ///
+        /// This requires passing in the hash of the child for validation
+        pub fn next(self, hash: Hash) -> AtBlobHeader {
+            let stream = ResponseDecoderStart::<TrackingReader<RecvStream>>::new(
+                hash.into(),
+                self.ranges,
+                IROH_BLOCK_SIZE,
+                self.reader,
+            );
+            AtBlobHeader {
+                stream,
+                misc: self.misc,
+            }
+        }
+
+        /// Finish the get response without reading further
+        pub fn finish(self) -> AtClosing {
+            AtClosing::new(self.misc, self.reader)
+        }
+    }
+
+    impl AtStartCollection {
+        /// The ranges we have requested for the child
+        pub fn ranges(&self) -> &RangeSet2<ChunkNum> {
+            &self.ranges
+        }
+
+        /// Go into the next state, reading the header
+        ///
+        /// For the collection we already know the hash, since it was part of the request
+        pub fn next(self) -> AtBlobHeader {
+            let stream = ResponseDecoderStart::new(
+                self.hash.into(),
+                self.ranges,
+                IROH_BLOCK_SIZE,
+                self.reader,
+            );
+            AtBlobHeader {
+                stream,
+                misc: self.misc,
+            }
+        }
+
+        /// Finish the get response without reading further
+        pub fn finish(self) -> AtClosing {
+            AtClosing::new(self.misc, self.reader)
+        }
+    }
+
+    /// State before reading a size header
+    #[derive(Debug)]
+    pub struct AtBlobHeader {
+        stream: ResponseDecoderStart<TrackingReader<RecvStream>>,
+        misc: Box<Misc>,
+    }
+
+    impl AtBlobHeader {
+        /// Read the size header, returning it and going into the `Content` state.
+        pub async fn next(self) -> Result<(AtBlobContent, u64), io::Error> {
+            let (stream, size) = self.stream.next().await?;
+            Ok((
+                AtBlobContent {
+                    stream,
+                    misc: self.misc,
+                },
+                size,
+            ))
+        }
+
+        /// Drain the response and throw away the result
+        pub async fn drain(self) -> result::Result<AtEndBlob, DecodeError> {
+            let (mut content, _size) = self.next().await?;
+            loop {
+                match content.next().await {
+                    BlobContentNext::More((content1, Ok(_))) => {
+                        content = content1;
                     }
+                    BlobContentNext::More((_, Err(e))) => {
+                        return Err(e);
+                    }
+                    BlobContentNext::Done(end) => {
+                        return Ok(end);
+                    }
+                }
+            }
+        }
+
+        /// Concatenate the response into a writer
+        pub async fn concatenate<W: AsyncWrite + Unpin, OW: FnMut(u64, usize)>(
```

Comment:
Not sure if that conveys the right meaning. This will concatenate the stuff it gets, even if there are gaps. I would expect write_into to take a SliceWriter and write things at the right position.

Reply:
I have updated the comment.